gh-119786: move interpreter doc from devguide to InternalDocs (#125715)
python/cpython: InternalDocs/interpreter.md

The bytecode interpreter
========================

Overview
--------

This document describes the workings and implementation of the bytecode
interpreter, the part of python that executes compiled Python code. Its
entry point is in
[Python/ceval.c](https://github.com/python/cpython/blob/main/Python/ceval.c).

At a high level, the interpreter consists of a loop that iterates over the
bytecode instructions, executing each of them via a switch statement that
has a case implementing each opcode. This switch statement is generated
from the instruction definitions in
[Python/bytecodes.c](https://github.com/python/cpython/blob/main/Python/bytecodes.c)
which are written in
[a DSL](https://github.com/python/cpython/blob/main/Tools/cases_generator/interpreter_definition.md)
developed for this purpose.

> Recall that the [Python Compiler](https://github.com/python/cpython/blob/main/InternalDocs/compiler.md)
> produces a [`CodeObject`](https://github.com/python/cpython/blob/main/InternalDocs/code_object.md),

Reviewer: I think you might be able to get away with just this (also applies to a bunch of other links in this PR):

```suggestion
Recall that the [Python Compiler](compiler.md) produces a [`CodeObject`](code_object.md),
```

Author: I guess I could also use relative links to the code (like `../Python/bytecodes.c`).

Reviewer: Try it and see how it renders? You can get a good preview of how GitHub will render it by pushing to this PR branch, then clicking this "View file" button here from the "files changed" tab:

---

![image](https://github.com/user-attachments/assets/ea3f0b5c-9844-4f92-b243-3b7192d63e12)

[`gen.throw`](https://docs.python.org/3.14/reference/expressions.html#generator.throw).

By default, [`_PyEval_EvalFrame()`](https://docs.python.org/3.14/c-api/veryhigh.html#c.PyEval_EvalCode)
simply calls [`_PyEval_EvalFrameDefault()`] to execute the frame. However, as per
[`PEP 523`](https://peps.python.org/pep-0523/) this is configurable by setting
`interp->eval_frame`. In the following, we describe the default function,
`_PyEval_EvalFrameDefault()`.

Instruction decoding
--------------------

The first task of the interpreter is to decode the bytecode instructions.
Bytecode is stored as an array of 16-bit code units (`_Py_CODEUNIT`).
Each code unit contains an 8-bit `opcode` and an 8-bit argument (`oparg`), both unsigned.
In order to make the bytecode format independent of the machine byte order when stored on disk,
`opcode` is always the first byte and `oparg` is always the second byte.
Macros are used to extract the `opcode` and `oparg` from a code unit
(`_Py_OPCODE(word)` and `_Py_OPARG(word)`).
Some instructions (for example, `NOP` or `POP_TOP`) have no argument -- in this case
we ignore `oparg`.

A simplified version of the interpreter's main loop looks like this:

> ```

Reviewer: This means that many renderers will provide syntax highlighting:

```suggestion
```c
```

    _Py_CODEUNIT *first_instr = code->co_code_adaptive;
    _Py_CODEUNIT *next_instr = first_instr;
    while (1) {
        _Py_CODEUNIT word = *next_instr++;
        unsigned char opcode = _Py_OPCODE(word);
        unsigned int oparg = _Py_OPARG(word);
        switch (opcode) {
        // ... A case for each opcode ...
        }
    }
```

This loop iterates over the instructions, decoding each into its `opcode`
and `oparg`, and then executes the switch case that implements this `opcode`.

The instruction format supports 256 different opcodes, which is sufficient.
However, it also limits `oparg` to 8-bit values, which is too restrictive.
To overcome this, the `EXTENDED_ARG` opcode allows us to prefix any instruction
with one or more additional data bytes, which combine into a larger oparg.

> For example, this sequence of code units::

Reviewer: Unlike ReST, the double-colon doesn't have any special meaning in Markdown:

```suggestion
For example, this sequence of code units:
```

    EXTENDED_ARG  1
    EXTENDED_ARG  0
    LOAD_CONST    2

would set `opcode` to `LOAD_CONST` and `oparg` to `65538` (that is, `0x1_00_02`).
The compiler should limit itself to at most three `EXTENDED_ARG` prefixes, to allow the
resulting `oparg` to fit in 32 bits, but the interpreter does not check this.

In the following, a `code unit` is always two bytes, while an `instruction` is a
sequence of code units consisting of zero to three `EXTENDED_ARG` opcodes followed by
a primary opcode.

The following loop, to be inserted just above the `switch` statement, will make the above
snippet decode a complete instruction:

> ```

Reviewer: ```suggestion
```c
```

    while (opcode == EXTENDED_ARG) {
        word = *next_instr++;
        opcode = _Py_OPCODE(word);
        oparg = (oparg << 8) | _Py_OPARG(word);
    }
```

For various reasons we'll get to later (mostly efficiency, given that `EXTENDED_ARG`
is rare) the actual code is different.

Jumps
=====

Note that when the `switch` statement is reached, `next_instr` (the "instruction offset")
already points to the next instruction.
Thus, jump instructions can be implemented by manipulating `next_instr`:

- A jump forward (`JUMP_FORWARD`) sets `next_instr += oparg`.
- A jump backward sets `next_instr -= oparg`.

Inline cache entries
====================

Some (specialized or specializable) instructions have an associated "inline cache".
The inline cache consists of one or more two-byte entries included in the bytecode

> array as additional words following the `opcode` /`oparg` pair.

Reviewer: ```suggestion
array as additional words following the `opcode`/`oparg` pair.
```

is created.

The stack effects of each instruction are also exposed through the
[opcode metadata](https://github.com/python/cpython/blob/main/Include/internal/pycore_opcode_metadata.h)
through two functions that report how many stack elements the instructions consumes,
and how many it produces (`_PyOpcode_num_popped` and `_PyOpcode_num_pushed`).
For example, the `BINARY_OP` instruction pops two objects from the stack and pushes the
result back onto the stack.

The stack grows up in memory; the operation `PUSH(x)` is equivalent to `*stack_pointer++ = x`,
whereas `x = POP()` means `x = *--stack_pointer`.
Overflow and underflow checks are active in debug mode, but are otherwise optimized away.

At any point during execution, the stack level is knowable based on the instruction pointer
alone, and some properties of each item on the stack are also known.
In particular, only a few instructions may push a `NULL` onto the stack, and the positions
that may be `NULL` are known.
A few other instructions (`GET_ITER`, `FOR_ITER`) push or pop an object that is known to
be an iterator.

Instruction sequences that do not allow statically knowing the stack depth are deemed illegal

> and the bytecode compiler never generates such sequences.

Reviewer: ```suggestion
Instruction sequences that do not allow statically knowing the stack depth are deemed illegal;
the bytecode compiler never generates such sequences.
```

> For example, the following sequence is illegal, because it keeps pushing items on the stack::

Reviewer: ```suggestion
For example, the following sequence is illegal, because it keeps pushing items on the stack:
```

builtins (though at runtime, the specializing interpreter can)
- cells: used for nonlocal references

(TODO: Write the rest of this section. Alas, the author got distracted and won't have time to continue this for a while.)

-->

<!--

Other topics
------------

(TODO: Each of the following probably deserves its own section.)

- co_consts, co_names, co_varnames, and their ilk
- How calls work (how args are transferred, return, exceptions)
- Eval breaker (interrupts, GIL)
- Tracing
- Setting the current lineno (debugger-induced jumps)
- Specialization, inline caches etc.

-->

Introducing a new bytecode instruction
--------------------------------------

It is occasionally necessary to add a new opcode in order to implement
a new feature or change the way that existing features are compiled.
This section describes the changes required to do this.

First, you must choose a name for the bytecode, implement it in
[`Python/bytecodes.c`](https://github.com/python/cpython/blob/main/Python/bytecodes.c)
and add a documentation entry in
[`Doc/library/dis.rst`](https://github.com/python/cpython/blob/main/Doc/library/dis.rst).
Then run `make regen-cases` to assign a number for it (see

> [`Include/opcode_ids.h`](https://github.com/python/cpython/blob/main/Include/opcode_ids.h)

Reviewer: unbalanced parentheses

```suggestion
Then run `make regen-cases` to assign a number for it (see
[`Include/opcode_ids.h`](https://github.com/python/cpython/blob/main/Include/opcode_ids.h))
```

and regenerate a number of files with the actual implementation of the bytecode in
[`Python/generated_cases.c.h`](https://github.com/python/cpython/blob/main/Python/generated_cases.c.h)
and metadata about it in additional files.

> With a new bytecode you must also change what is called the magic number for

Reviewer: ```suggestion
With a new bytecode you must also change what is called the "magic number" for
```

Changing this number will lead to all .pyc files with the old `MAGIC_NUMBER`
to be recompiled by the interpreter on import.  Whenever `MAGIC_NUMBER` is
changed, the ranges in the `magic_values` array in
[`PC/launcher.c`](https://github.com/python/cpython/blob/main/PC/launcher.c)
may also need to be updated.  Changes to
[`Lib/importlib/_bootstrap_external.py`](https://github.com/python/cpython/blob/main/Lib/importlib/_bootstrap_external.py)
will take effect only after running `make regen-importlib`.

> [!NOTE]
> Running `make regen-importlib` before adding the new bytecode target to
> [`Python/bytecodes.c`](https://github.com/python/cpython/blob/main/Python/bytecodes.c)
> (followed by `make regen-cases`) will result in an error. You should only run
> `make regen-importlib` after the new bytecode target has been added.

> [!NOTE]
> On Windows, running the `./build.bat` script will automatically
> regenerate the required files without requiring additional arguments.

Finally, you need to introduce the use of the new bytecode.  Update
[`Python/codegen.c`](https://github.com/python/cpython/blob/main/Python/codegen.c)

> to emit code with this bytecode. Optimizations in :cpy-file:`Python/flowgraph.c`

Reviewer: I don't think the :cpy-file: role has any effect in a Markdown file; it only works in ReST
